### DISCUSSION TITLE
Plugin UUID sent to registerPlugin is not persistent

### DISCUSSION THREAD

**Discussion by JimThatcher** [elgatosf/elgato#53] [Help] [answered]
Every time Stream Deck restarts (computer restarts, or I exit Stream Deck to update plugin code) the UUID that Stream Deck sends in the registerPlugin process is different. Since this value is a parameter for calling get/setGlobalSettings this means that global settings are not persisted across app sessions, rendering them useless.

My workaround for this is ugly. It appears that action-specific settings are actually persisted across sessions, so I put a copy of the plugin global settings in the settings for every action to ensure that those values will be available when the actions appear and are executed. This works, but it means I am storing dozens of copies of the same settings.

Please fix the Stream Deck app so that get/setGlobalSettings provides a persistent storage for plugins' global settings.

**Comment by tiptronic** [comment, marked as answer]
Hi @JimThatcher 

`setGlobalSettings` uses the UUID of your plugin which is a unique value during the lifetime of the plugin's instance and is used _internally_ by StreamDeck to identify the plugin. The value is created and mapped (internally) every time the plugin's instance is created. 

So, using `getGlobalSettings` and  `setGlobalSettings` is the perfect way to use settings for all your action.

I can't say what's going on on your end, but the events work just fine.
E.g. If you use our [library ](https://github.com/elgatosf/streamdeck-javascript-sdk) you can quickly check it is working properly:

```javascript
$SD.onDidReceiveGlobalSettings((jsn) => {
    // the log should show the `lastchecked` value sent to StreamDeck (see below)
    console.log('onDidReceiveGlobalSettings', jsn);
});

$SD.onConnected(jsn => {
    console.log('Property Inspector connected', jsn);
    // important is to call the 'get/setGlobalSettings' events
    // AFTER the plugin is connected

    $SD.getGlobalSettings();
    $SD.setGlobalSettings({lastchecked: new Date().toLocaleString()});
});
```

In the above example, `set/getGlobalSettings` is called like this:

### getGlobalSettings
```json
{
    "context": "12347BADD02930ABBE997B9DD7348027",
    "event": "getGlobalSettings"
}
```

### setGlobalSettings
```json
{
    "context": "12347BADD02930ABBE997B9DD7348027",
    "event": "setGlobalSettings",
    "payload": {
        "lastchecked": "09/08/2023, 09:39:31"
    }
}
```

If you don't use our library, you can try this:

```javascript
// replace the 'context' value with your plugin's uuid (see below)
// make sure you put everything in quotes
// and then call it using the WebSocket connection opened when starting the communication
yourwebsocket.send('{"context":"12347BADD02930ABBE997B9DD7348027","event":"getGlobalSettings"}'); 
```

You will get the `uuid` as the second argument in `connectElgatoStreamDeckSocket`
```javascript
function connectElgatoStreamDeckSocket(port, uuid, messageType, appInfoString, actionInfo)
```
in turn you will get the `didReceiveSettings` event with the previously set value as payload.
This works until you remove your plugin from Stream Deck.

Important is only that you call `setGlobalSettings` AFTER your plugin is registered and connected to StreamDeck.

If it doesn't work for you, please show your code so we can see what's going wrong there.
